File: Java/programmers/etc/level4/사칙연산.java
package CodeTest.Java.programmers.etc.level4;

import java.util.*;

public class 사칙연산 {
    public int solution(String[] arr) {
        int n = (arr.length+1)/2;

        int[] num = new int[n];
        char[] op = new char[n-1];

        for (int i=0; i<n; i++) num[i] = Integer.valueOf(arr[i*2]);
        for (int i=0; i<n-1; i++) op[i] = arr[i*2+1].charAt(0);

        int[][] MIN_DP = new int[n][n];
        for (int[] row : MIN_DP) Arrays.fill(row, Integer.MAX_VALUE);

        int[][] MAX_DP = new int[n][n];
        for (int[] row : MAX_DP) Arrays.fill(row, Integer.MIN_VALUE);

        for (int step=0; step<n; step++) {
            for (int i=0; i<n-step; i++) {
                int j = i+step;
                if (step == 0) {
                    MIN_DP[i][i] = num[i];
                    MAX_DP[i][i] = num[i];
                }
                else {
                    for (int k=i; k<j; k++) {
                        if (op[k] == '+') {
                            MAX_DP[i][j] = Math.max(MAX_DP[i][j], MAX_DP[i][k]+MAX_DP[k+1][j]);
                            MIN_DP[i][j] = Math.min(MIN_DP[i][j], MIN_DP[i][k]+MIN_DP[k+1][j]);
                        }
                        else {
                            MAX_DP[i][j] = Math.max(MAX_DP[i][j], MAX_DP[i][k]-MIN_DP[k+1][j]);
                            MIN_DP[i][j] = Math.min(MIN_DP[i][j], MIN_DP[i][k]-MAX_DP[k+1][j]);
                        }
                    }
                }
            }
        }

        return MAX_DP[0][n-1];
    }
}

File: Python/programmers/etc/level4/사칙연산.py
import sys

def solution(arr):
    INF = sys.maxsize
    
    num, op = [], []
    for a in arr:
        if a.isnumeric():
            num.append(int(a))
        else:
            op.append(a)
    
    n = len(num)
    MIN_DP = [[INF]*n for _ in range(n)]
    MAX_DP = [[-INF]*n for _ in range(n)]

    for step in range(n):
        for i in range(n-step):
            j = i+step
            if step == 0:
                MIN_DP[i][i] = num[i]
                MAX_DP[i][i] = num[i]
            else:
                for k in range(i, j):
                    if op[k] == "+":
                        MAX_DP[i][j] = max(MAX_DP[i][j], MAX_DP[i][k]+MAX_DP[k+1][j])
                        MIN_DP[i][j] = min(MIN_DP[i][j], MIN_DP[i][k]+MIN_DP[k+1][j])
                    else:
                        MAX_DP[i][j] = max(MAX_DP[i][j], MAX_DP[i][k]-MIN_DP[k+1][j])
                        MIN_DP[i][j] = min(MIN_DP[i][j], MIN_DP[i][k]-MAX_DP[k+1][j])

    return MAX_DP[0][n-1]


arr = ["1", "-", "3", "+", "5", "-", "8"]
print(solution(arr)) # 1

arr = ["5", "-", "3", "+", "1", "+", "2", "-", "4"]
print(solution(arr)) # 3
